design/authenticators/authn_sut/authn_sut_solution_design.md

```diff
@@ -0,0 +1,409 @@
+| 11 | Missing code challenge algorithm in /login request headers | Code challenge algorithm not provided    |
+| 12 | Missing code verifier in /authenticate request body | Code verifier not provided    |
+| 13 | Code verifier does not match code challenge | Code verifier does not match code challenge    |
+
+###### Debug Log level
+
+|    | Scenario                                              | Log message                                                                            |
+|--- |-------------------------------------------------------|----------------------------------------------------------------------------------------|
+| 1  | After validating SUT | Single use token for {0-role-name} has been validated successfully                |
+
+##### Audit Log Messages
+
+Same as other authenticators - one message for successful authentication, one for failed authentication.
+
+#### Policy
+
+As with other authenticators, the SUT authenticator will be disabled by default.
+It will be enabled by creating an authenticator policy resource in Conjur and
+adding it to the list of enabled authenticators in the Conjur configuration, for
+example in the `CONJUR_AUTHENTICATORS` environment variable.
+
+The authenticator webservice must be declared in Conjur policy:
+
+```yml
+- !policy
+  id: conjur/authn-sut
+  body:
+  - !webservice
+
+  - !group 
+    id: authenticatable
+    annotations:
+      description: Group with permission to authenticate using this authenticator
+
+  - !permit
+    role: !group authenticatable
+    privilege: [ read, authenticate ]
+    resource: !webservice
+```
+
+and the `authenticatable` group can be used to entitle a user(s) to use the
+authenticator:
+
+```yml
+- !grant
+  role: !group conjur/authn-sut/authenticatable
+  member: !user <username>
+```
+
+#### API
+
+We will need two new API endpoints for the SUT authenticator:
+
+##### `POST /authn-sut/<account>/login`
+
+This endpoint will be used to generate a SUT for a user. The user will provide
+their username and password (or API key) in the `Authentication` header using the
+standard format for HTTP Basic authentication.
+
+Additionally, the client will provide a Code Challenge in the request body and a Code Challenge Algorithm
+in the headers. For example:
+
+Headers:
+
+```txt
+Authorization: Basic ...
+Code-Challenge-Algorithm: sha256
+```
+
+Body:
+
+```json
+{
+  "code_challenge": "3iod..."
+}
+```
+
+For more details on the code challenge, see the [Code Verifier and Code Challenge](#code-verifier-and-code-challenge) section.
+
+The endpoint will return a SUT that can be used to authenticate the user to Conjur.
+
+The format of the URL matches the existing authenticator endpoints for authn and
+authn-ldap. However, the request action for authn-sut is `POST` instead of `GET`
+since we need to send a code challenge in the request body.
+
+##### `POST /authn-sut/<account>/authenticate`
```

Comment:
As I was thinking more about the interface of the other authenticators, I realized they typically include the role ID: 

- `POST /authn-jwt/{service-id}/<account>/[{host-id}]/authenticate`
- `POST /authn-azure/{service-id}/{account}/{host-id}/authenticate`
- `POST /{authenticator}/{account}/{login}/authenticate`

I think it would be valuable to follow this convention and require the role identifier (host or username) in the URI.

Reply:
Hmm. We don't use this format for authn or authn-ldap, which are the only two authenticators that also have a /login route and what I've been mostly basing this on. Perhaps it's because the username is part of the HTTP Basic auth header.

Reply:
Not login, but the default authenticator (role and API key) uses this pattern: https://docs.cyberark.com/conjur-enterprise/latest/en/Content/Developer/Conjur_API_Authenticate.htm?tocpath=Developer%7CREST%C2%A0APIs%7C_____3

Reply:
Ah I see, I missed that nuance. I'll update this.

Reply:
I'd also make the argument that mandating the role identifier in the URI makes the implementation slightly more secure by forcing the requestor to know the role and the role's single-use token before the request. If the provided role identifier and the single-use token role don't match, the authentication should fail.

Reply:
Good point. I've updated this.